### ISSUE TITLE
Vine: Explicit Limit on FunctionCall Threads

### ISSUE THREAD

**Issue by dthain** [cooperative-computing-lab/cctools#3734]
Observed when running Python Numpy codes at scale on the CRC:

A LibraryTask is declared to use`N` threads and `N` slots, with the intention of running multiple FunctionCall tasks concurrently.  Vine (correctly) sets the environment variable `OMP_NUM_THREADS` for the LibraryTask before it runs, to inform it of the correct number of threads.  The LibraryTask then forks `N` concurrent FunctionCalls.  So far so good.

However, the problem is that each FunctionCall task inherits the `OMP_NUM_THREADS` setting, and then **all** of them use `N` threads simultaneously.  This results in a total of `N*N` threads running on the machine.  Yikes!

What should happen instead is that the LibraryTask should set `OMP_NUM_THREADS` (and related variables) to the total number of threads divided by the number of function slots.  Then we should get the proper number of threads running concurrently.

**Comment by JinZhou5042** [comment]
One thing comes to my mind is that the LibraryTask itself is a process, so should a LibraryTask come with `N` FunctionCall tasks create `N+1` processes in total? If so, 23 FunctionCalls and 1 LibraryTask would be an ideal configuration for a 24-core worker?

**Comment by dthain** [comment]
Well, it depends on whether the LibraryTask parent process is going to do a significant amount of computation.  If all it does is fork children and collect the results, I would argue no.  If it is going to do something more complex, then maybe yes.

**Comment by JinZhou5042** [comment]
Explicitly setting `OMP_NUM_THREADS` seems not help mitigate per-task CPU load.

Here is the way each task sets `OMP_NUM_THREADS` and in my case it is 1.
<img width="802" alt="image" src="https://github.com/cooperative-computing-lab/cctools/assets/142265839/3a99ac8a-2ea6-4d28-b8d0-ba2b3cefe64e">


However, while doing computation expensive tasks, CPU load on each task is still nearly 300%, which is roughly the same than before.
<img width="739" alt="image" src="https://github.com/cooperative-computing-lab/cctools/assets/142265839/baf002e2-d28f-4bd9-8984-7c5cb8f61bd6">

I tested the cpu load over a single function locally and it is continuously about 100%.

**Comment by JinZhou5042** [comment]
<img width="1639" alt="image" src="https://github.com/cooperative-computing-lab/cctools/assets/142265839/c4f24d65-cc2a-4d1f-b8c7-766def80f49d">

**Comment by dthain** [comment]
Hmm, so you certain that the child process receives the correct environment variable?  

The other possibility is that the code in the function doesn’t respect those specific variables, and there is some other way to control it.

Is the function using numpy or something else?

**Comment by JinZhou5042** [comment]
Yes, the return value of each function is exactly the env variable `OMP_NUM_THREADS` and it is 1. 
Those functions are using number, but I did an experiment which forks a subprocess to run that function from the main, it turns out to be using 1 core.
I am looking into if there are some other env variables affect the forked processes.

**Comment by JinZhou5042** [comment]
You are right, numpy uses as many cores as it can. As what I have understood, those environment variables should be set before importing numpy and therefore it doesn't work if just set them before forking.

**Comment by JinZhou5042** [comment]
I now believe you that numpy uses multiple cores in a forked process. I did three more experiments, each one has 200 tasks, over 1 worker with 8 cores. By setting the cores of library task to 1, 2, 4 and 8 and the function slots equaling to cores (which means the `OMP_NUM_THREADS` always equals to 1), the average CPU load on each allocated core has the following evolution:

<img width="1044" alt="image" src="https://github.com/cooperative-computing-lab/cctools/assets/142265839/07bffcde-547c-49f0-b0ad-0d3bf5a6b78f">

<img width="1106" alt="image" src="https://github.com/cooperative-computing-lab/cctools/assets/142265839/ab1b3f4e-b8d0-4273-b177-7c711f92ff80">

<img width="1173" alt="image" src="https://github.com/cooperative-computing-lab/cctools/assets/142265839/ab7dc2ba-a616-4660-879a-6ba5651d098e">

<img width="1120" alt="image" src="https://github.com/cooperative-computing-lab/cctools/assets/142265839/67c19561-7734-4708-a9f7-4a3a79cb90c4">


I think numpy uses as many available cores as possible to accelerate the computation. At a certain time, a single core may simultaneously execute matrix multiplication operations for different tasks, so that the average execution time and the gap between tasks are notably reduced.

Setting environment variables are seemingly useless, and I am still confused why those functions running on my local machine only consume 1 core ...

**Comment by JinZhou5042** [comment]
That is to say, libraries with only 1 core have the fast speed in terms of average function execution time.

**Comment by JinZhou5042** [comment]
Functions without numpy perform well with reasonable CPU load per task, this one uses 4 lib cores and 2 func slots:
<img width="559" alt="image" src="https://github.com/cooperative-computing-lab/cctools/assets/142265839/fcfd025f-8b31-4ddc-a357-30032b87e63f">
And this one has 4 lib cores and 4 func slots:
<img width="553" alt="image" src="https://github.com/cooperative-computing-lab/cctools/assets/142265839/c3e8fb20-c80e-4fee-aaab-482e7d0f27b8">


For functions using numpy, things are quite complex, with 8 cores available per worker, this one uses 1 lib core and 1 func slot:
<img width="645" alt="image" src="https://github.com/cooperative-computing-lab/cctools/assets/142265839/c98bede4-b64f-4274-a8ef-0034ade622d7">

This one has 2 lib cores and 2 func slots:
<img width="636" alt="image" src="https://github.com/cooperative-computing-lab/cctools/assets/142265839/324dbc43-4b51-44c3-94dc-a16946987391">

This one has 4 lib cores and 4 func slots:
<img width="646" alt="image" src="https://github.com/cooperative-computing-lab/cctools/assets/142265839/9cbefd86-aefc-4ccf-9c19-ba271e75b0ec">

This one has 8 lib cores and 8 func sllots:
<img width="649" alt="image" src="https://github.com/cooperative-computing-lab/cctools/assets/142265839/68a0622f-5c25-49ec-a7c1-213f6fcd4a78">

**Comment by JinZhou5042** [comment]
I am thinking can we dynamically increase or decrease the following configurations:
- number of workers
- number of total cores, memory, disk on each worker
- number of library cores and function slots per function

**Comment by dthain** [comment]
Let's not get complicated until we understand the basic configuration well.

Does it matter if you set the environment variables prior to importing numpy?

This would seem to be relevant:
https://stackoverflow.com/questions/30791550/limit-number-of-threads-in-numpy

**Comment by JinZhou5042** [comment]
Setting environment variables prior to importing numpy seems not to matter, if numpy is able to use multiple cores to accelerate and we can exactly control the resources to each function, then it is critical to allocate the right number of cores to improve overall performance. 

I am investigating how to control the usable cores for each function using numpy.

**Comment by JinZhou5042** [comment]
The module [threadpoolctl](https://github.com/joblib/threadpoolctl) provided by @btovar manages threads used for numpy gracefully, each task now consumes the exact number of cores as given. 

<img width="750" alt="image" src="https://github.com/cooperative-computing-lab/cctools/assets/142265839/205a8799-5796-455e-81e8-5f630beb66c5">

As it is a built-in package in Python, I am thinking can we directly use it inside of the function wrapper in the library network code? What do you think @dthain

**Comment by btovar** [comment]
I don't think it is a core python module, though. If it does not bring big dependencies we could add it as requirement for ndcctools in conda.

**Comment by dthain** [comment]
@JinZhou5042 , what do you mean by "frustrated", is the performance markedly different than before?

Also, can you briefly explain how this library works?  I would like to understand better whether we should just import it and use it (which has some cost in dependencies) or whether we should implement its techniques ourselves.

**Comment by JinZhou5042** [comment]
@dthain with all other configurations keep the same, same worker, same cores, memory and function along with the arguments, the one using threadpoolctl for resource management increases the total execution time from 223s to 678s, with the average function runtime growing from 8.8s to 27s

**Comment by dthain** [comment]
What machine? How many cores?  How many threads?

**Comment by JinZhou5042** [comment]
1 machine which is d12chas328.crc.nd.edu, 8 cores in total but I only allocate 4 cores to functions with each function operating on a single core. The comparison can be seen in the archived file.

[Archive.zip](https://github.com/cooperative-computing-lab/cctools/files/14923394/Archive.zip)

**Comment by dthain** [comment]
So consider this:

In the uncontrolled case, your functions are going outside the bounds of what they are allocated, effectively stealing cores from other uses.  The machine has 12 cores, and each of your four function instances are trying to use all 12 cores. In the controlled case, your functions are "behaving" and each of the four is using one core.

So, of course the uncontrolled case runs faster, because your functions are "overflowing" their allocation and using cores that they shouldn't!

If you want to make a more fair comparison, then start a 12 core worker on a 12 core machine to run 12 functions.  In the uncontrolled case, you will get 144 threads competing for 12 cores.  In the controlled case, you will get 12 threads running on 12 workers.  And there I suspect the controlled case will run faster because it is not overcommitted.

**Comment by JinZhou5042** [comment]
Yes, you are totally correct! I tested those functions on 4 24-core workers and made the library operating on all of the available cores, it turns out that the controlled case has better performance including the average execution time and overall runtime. I suspect it is because the uncontrolled case starts a bunch of threads for competing extra cores and that was in vain because all cores are consumed by the library.

I archived the reports in case you are interested to see.

[Archive.zip](https://github.com/cooperative-computing-lab/cctools/files/14927674/Archive.zip)
